### PR TITLE
Introduced custom animal species classification mode

### PR DESCRIPTION
**Pull request by stefanodellosa-personal**
This allows to enable a custom classification mode based on parametric species (selected by user) on which base the notification. 

It basically is a generalisation of bear detection mode.

### CODE REVIEW

File: wadas/ui/mainwindow.py [hunk 1]
```diff
@@ -196,17 +197,27 @@ def run(self):
...
-                    return
-                OperationMode.cur_operation_mode.url = self.url_input_dialog()
-                if not OperationMode.cur_operation_mode.url:
-                    logger.error("Cannot proceed without a valid URL. Please run again.")
-                    return
-            elif not cameras:
+            match OperationMode.cur_operation_mode.type:
+                case OperationMode.OperationModeTypes.TestModelMode:
+                    OperationMode.cur_operation_mode.url = self.url_input_dialog()
+                    if not OperationMode.cur_operation_mode.url:
```
Comment (cdimauro) [inline wadas/ui/mainwindow.py:205]: The := operator can be used.

File: wadas/ui/mainwindow.py [hunk 2]
```diff
@@ -196,17 +197,27 @@ def run(self):
...
+                        logger.error("Cannot proceed without a valid URL. Please run again.")
+                        return
+                case OperationMode.OperationModeTypes.CustomSpeciesClassificationMode:
+                    selected_species = self.custom_species_dialog()
+                    if selected_species:
+                        OperationMode.cur_operation_mode.target_animal_label = selected_species
```
Comment (cdimauro) [inline wadas/ui/mainwindow.py:211]: Same here.

File: wadas/ui/mainwindow.py [hunk 3]
```diff
@@ -362,6 +376,17 @@ def url_input_dialog(self):
             logger.warning("URL insertion aborted.")
             return ""
 
+    def custom_species_dialog(self):
+        """Method to run a dialog to return selected custom species to classify."""
+
+        species_dialog = DialogSelectAnimalSpecies()
+        if species_dialog.exec_():
```
Comment (cdimauro) [inline wadas/ui/mainwindow.py:383]: :=

File: wadas/ui/select_animal_species.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
+import os
+
+from PySide6.QtGui import QIcon
+from PySide6.QtWidgets import QDialog
+
+from wadas.ai.models import txt_animalclasses
+from wadas.ui.qt.ui_select_animal_species import Ui_DialogSelectAnimalSpecies
+
+module_dir_path = os.path.dirname(os.path.abspath(__file__))
+
```
Comment (cdimauro) [inline wadas/ui/select_animal_species.py:10]: Add another blank line.

File: wadas/ui/select_mode_dialog.py [hunk 5]
```diff
@@ -43,6 +43,8 @@ def initialize_radiobutton_selection(self):
                 self.ui.radioButton_tunnel_mode.setChecked(True)
             elif OperationMode.cur_operation_mode_type == OperationMode.OperationModeTypes.BearDetectionMode:
                 self.ui.radioButton_bear_det_mode.setChecked(True)
+            elif OperationMode.cur_operation_mode_type == OperationMode.OperationModeTypes.CustomSpeciesClassificationMode:
+                self.ui.radioButton_custom_species_class_mode.setChecked(True)
```
Comment (cdimauro) [inline wadas/ui/select_mode_dialog.py:47]: Maybe the match/case construct could be used to replace this chain of if/elifs.

File: wadas/ui/select_mode_dialog.py [hunk 6]
```diff
@@ -55,4 +57,6 @@ def accept_and_close(self):
             OperationMode.cur_operation_mode_type = OperationMode.OperationModeTypes.AnimalDetectionAndClassificationMode
         elif self.ui.radioButton_bear_det_mode.isChecked():
             OperationMode.cur_operation_mode_type = OperationMode.OperationModeTypes.BearDetectionMode
+        elif self.ui.radioButton_custom_species_class_mode.isChecked():
+            OperationMode.cur_operation_mode_type = OperationMode.OperationModeTypes.CustomSpeciesClassificationMode
```
Comment (cdimauro) [inline wadas/ui/select_mode_dialog.py:61]: Same here: match/case?

Comment (stefanodellosa-personal) [reply wadas/ui/select_mode_dialog.py:61]: not really, are different objects to check. not sure how to automate in match case

File: wadas/ui/mainwindow.py (hunk 3 above)
Comment (stefanodellosa-personal) [reply wadas/ui/mainwindow.py:383]: not sure how to create object and call its method on assigment.

File: wadas/ui/select_mode_dialog.py (hunk 6 above)
Comment (cdimauro) [reply wadas/ui/select_mode_dialog.py:59]: Yes, it's a bit tricky. It requires a bit of redesign to achieve it, but it's not the right moment now.

File: wadas/ui/mainwindow.py (hunk 3 above)
Comment (cdimauro) [reply wadas/ui/mainwindow.py:382]: if (species_dialog := DialogSelectAnimalSpecies()).exec_():

**Pull request merged**